### ISSUE TITLE
File locking support for better game compatibility (SHARE.EXE)

### ISSUE THREAD

**Issue by Grounded0** [dosbox-staging/dosbox-staging#1489]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### If Other OS, please describe

macOS 11.6.2/x86-64

### Is your feature request related to a problem? Please describe.

When installing some of the more demanding Windows games this pops up:

<img width="1392" alt="Screenshot 2022-09-07 at 17 28 27" src="https://user-images.githubusercontent.com/48192273/188904227-6ffb0af9-ce88-449e-a5f6-16708170025f.png">

This is due to lack of file locking support in DOSBox Staging which affects how both SHARE and VSHARE.386 do file locking.

At least the following games are known to suffer from SHARE.EXE not loaded issue:

- Comix Zone (1995)
- Garfield Caught in the Act (1996)
- Microsoft Flight Simulator 5 - ATC Workshop (1995)
- Monopoly (1995)
- Ultimate Yahtzee (1996)

### Describe the solution you'd like

I would have previously opposed adding file locking support as a too complex to implement. We've since moved to C++17 which has filesystem library that makes this feature easier to implement and more robust to maintain. The best way to implement is by accessing host OS directory mount directly.

The file locking logic can be pretty easily summed up by a snippet of code from DOSBox-X that has this implemented:

dos_files.cpp:890:

```
	} else {
		//Test if file exists, but opened in read-write mode (and writeprotected)
		if((((flags&3) != OPEN_READ) || (enable_share_exe && !strncmp(Drives[drive]->GetInfo(),"local directory ",16))) && Drives[drive]->FileExists(fullname))
			DOS_SetError(DOSERR_ACCESS_DENIED);
		else {
			if(!PathExists(name)) DOS_SetError(DOSERR_PATH_NOT_FOUND); 
			else DOS_SetError(DOSERR_FILE_NOT_FOUND);
		}
		return false;
	}
```

Adding this feature with the upcoming DOS 38h support will make this a general purpose emulator for DOS and Win16 environments. This makes it easier to use for a basic user because running multiple emulators for the same environment can be avoided. This will also attract more users.

### Describe alternatives you've considered

There is a method of using fake SHARE but such solutions may lead to data corruption.

The question of is this in the scope of the project arises in this case. My evaluation is that it is based on the Vision page:

https://dosbox-staging.github.io/about/

<img width="601" alt="Screenshot 2022-01-02 at 15 11 23" src="https://user-images.githubusercontent.com/48192273/147877030-194980d4-d50e-41cc-9454-348300ab33ad.png">

By utilising C++17 filesystem in implementing i think it passes all the conditions to be a viable, implementable feature. Can be implemented after 0.79 release cycle if there's no resources.

### Add any other context or screenshots about the feature request here.

Sample code for C++17 filesystem file locking:

https://stackoverflow.com/a/56348495

```
#include <sys/stat.h>

bool IsLocked(std::string& thePath)
{
    bool isLocked = false;

    struct stat buf;
    int retval = stat(thePath.c_str(), &buf);     // NOTE: retval could be an error, like ENOENT (file doesn't exist)
    if (0 != (buf.st_flags & UF_IMMUTABLE)) {
        isLocked = true;
         //is immutable
    }

    return isLocked;
}
```

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by Grounded0** [comment]
![Screenshot_20220102-212202](https://user-images.githubusercontent.com/48192273/147887278-2b84e1c0-793e-44c0-a6ee-9dc4590c94fc.png)

![Screenshot_20220102-212301](https://user-images.githubusercontent.com/48192273/147887295-6d2dda82-7b1f-4dfc-9a3b-bbb8d4f9d47c.png)

![Screenshot_20220102-212601](https://user-images.githubusercontent.com/48192273/147887298-3c765fa4-670b-4ccc-8bfc-f34fc1554519.png)

![Screenshot_20220102-212627](https://user-images.githubusercontent.com/48192273/147887303-2ac3309a-25fd-4e6e-b414-b4cb1af7b8b7.png)

**Comment by Grounded0** [comment]
<img width="721" alt="Screenshot 2022-09-08 at 19 39 23" src="https://user-images.githubusercontent.com/48192273/189178254-eaf21f87-77e8-43e2-b28d-77548d867fcd.png">
<img width="721" alt="Screenshot 2022-09-08 at 19 39 46" src="https://user-images.githubusercontent.com/48192273/189178270-600857a7-3652-468f-8109-4bbdf61ac729.png">
<img width="721" alt="Screenshot 2022-09-08 at 19 40 00" src="https://user-images.githubusercontent.com/48192273/189178276-1e0e9b04-92e2-4edd-bf5c-6535c140f983.png">

**Comment by kcgen** [comment]
In addition to productivity applications, Rich **_¥Weeds¥_** Nagel [mentioned](https://www.vogons.org/memberlist.php?mode=viewprofile&u=16870) another gaming use for SHARE:

> Microsoft Flight Simulator 5, the ATC Workshop that comes with the FS5 addon Flight Shop. ATC workshop allows you to create flight "adventures" for FS5 that are totally air traffic controlled.

> When I run the utility (FSFSADV.EXE) in Windows 3.11 under DOSBox I get an error message stating the SHARE.EXE hasn't been loaded (see attached screenshot), and then the utility immediately closes/shuts down. The readme for ATC Workshop states that:

``` text
WINDOWS 3.1 OWNERS
------------------
The ATC Workshop component of Flight Shop requires SHARE.EXE to be in memory
and configured properly to function. You will need to edit your CONFIG.SYS
file and add or alter the following line:
    
INSTALL=C:\DOS\SHARE.EXE /L:25
    
You will not be able to use ATC Workshop without this addition to your 
CONFIG.SYS.
    
If you are using Windows for Workgroups 3.11 or Windows 95 this is not
an issue.
```

> I dug up a copy from my old backups of SHARE.EXE for MS-DOS 6.22 and ran it before starting Windows 3.11 under DOSBox, but get the same error message. Any idea of a workaround for this?

> I recently installed my old copy of Microsoft Flight Simulator 2000 Professional, and have been having a blast flying it. FS2000 supports adventures files from the much older Flight Shop that have been converted to the newer FS2000 format (with the AAFConv.exe utility that comes with FS2000), and I thought that it would be fun using some newly created ones. Hoping that I still can 😀

![SHARE EXE_Error](https://user-images.githubusercontent.com/1557255/198894892-37c37796-0d66-48cb-b633-0755fa36a8be.png)

Post is Copyright "Public domain"; all credit to Rich and Vogons.

**Comment by nemo93** [comment]
The following games are known to suffer the dreaded issue "share.exe not loaded" during install:

* Comix Zone (SEGA)
* Garfield Caught in the Act (SEGA)
* Microsoft Flight Simulator 5 - ATC Workshop (Microsoft)
* Monopoly
* Ultimate Yahtzee (Hasbro)

**Comment by johnnovak** [comment]
I'd say this is a good addition as it improves compatibility. Also, the issue can be rightfully seen as a gap in the emulation.

Given it seems it can be done cleanly, and it's a small change, it's a no brainer, isn't it? I don't see any drawbacks, only benefits.

**Comment by johnnovak** [comment]
For reference, we've documented the dodgy and less-than-ideal `FAKESHAR.COM` solution on our Wiki:

https://github.com/dosbox-staging/dosbox-staging/wiki/Windows-3.1x#shareexe-not-loaded

**Comment by Python-Exoproject** [comment]
Games eXoWin3x has using fakeshar.com (in addition to games mentioned previously in this ticket) are:

3-D Ultra Pinball (1995)
AFL Finals Fever (1996)
Macbeth (1995)
Microsoft Cinema '95 (1995)
Mind Grind (1996)
RollOver (1997)

**Comment by weirddan455** [comment]
I'll be working on this in the coming days.  I spend some time looking at implementations of MS-DOS 4.0 source code dump, FreeDOS, and DOSBox-X.  `SHARE.EXE` is a program that is part of MS-DOS but not Windows.

My initial findings are that getting MS-DOS's `SHARE.EXE` to run under DOSBox would be out of scope.  `SHARE.EXE` runs as a TSR that installs interrupt handlers into core file handling routines in INT 21H.  Instead, we can emulate it's behavior in our internal code, similar to what DOSBox-X has done.

Unlike DOSBox-X's approach which depends on host OS functionality such as the Win32 API function `LockFile`, I am planning to implement internal file locking behavior that can be applied to any arbitrary `DOS_OpenFile` call.  This should allow file locking to work regardless of the underlying OS or file system and without having to deal with OS-specific behavior.

**Comment by johnnovak** [comment]
Very nice @weirddan455, looking forward to this 🚀 

I will help test all the games in this ticket, including the affected eXoWin3.x games listed by [Python-Exoproject](https://github.com/Python-Exoproject) when the time comes.

**Comment by johnnovak** [comment]
> Games eXoWin3x has using fakeshar.com (in addition to games mentioned previously in this ticket) are:
> 
> 3-D Ultra Pinball (1995) AFL Finals Fever (1996) Macbeth (1995) Microsoft Cinema '95 (1995) Mind Grind (1996) RollOver (1997)

@Python-Exoproject In theory, this is all that's needed to get the games on your list running. I'm intending to test these in the coming weeks.
